components/core/src/clp/hash_utils.hpp

```diff
@@ -0,0 +1,122 @@
+     * @throw EvpCtxManager::ErrorCode_NotInit if EVP_DigestInit_ex fails.
+     */
+    EvpDigestContext(EVP_MD const* type) : m_md_ctx{EVP_MD_CTX_create()} {
+        if (nullptr == m_md_ctx) {
+            throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+        }
+        // Set impl to nullptr to use the default implementation of digest type
+        if (1 != EVP_DigestInit_ex(m_md_ctx, type, nullptr)) {
+            throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+        }
+    }
+
+    // Disable copy constructor/assignment operator
+    EvpDigestContext(EvpDigestContext const&) = delete;
+    auto operator=(EvpDigestContext const&) -> EvpDigestContext& = delete;
+
+    // disable move constructor/assignment operator
+    EvpDigestContext(EvpDigestContext&&) = delete;
+    auto operator=(EvpDigestContext&&) -> EvpDigestContext& = delete;
+
+    // Destructor
+    ~EvpDigestContext() { EVP_MD_CTX_destroy(m_md_ctx); }
+
+    // Methods
+    /**
+     * Calls EVP_DigestUpdate to digest the input and update the hash context
+     * @param input
+     * @return ErrorCode_Success on success.
+     * @return ErrorCode_Unsupported if context is already finalized.
+     * @return ErrorCode_Failure if EVP_DigestUpdate fails.
+     */
+    [[nodiscard]] auto digest_update(std::span<unsigned char const> input) -> ErrorCode;
+
+    /**
+     * Calls EVP_DigestFinal_ex to generate the final hash result
+     * @param hash Returns the hashing result.
+     * @return ErrorCode_Success on success.
+     * @return ErrorCode_Unsupported if context is already finalized.
+     * @return ErrorCode_Corrupt if the hashing result has an unexpected length.
+     * @return ErrorCode_Failure if EVP_DigestFinal_ex fails.
+     */
+    [[nodiscard]] auto digest_final_ex(std::vector<unsigned char>& hash) -> ErrorCode;
+
+private:
+    EVP_MD_CTX* m_md_ctx{nullptr};
+    bool m_is_digest_finalized{false};
+};
+
+/**
+ * Converts input hash into a hex string.
+ * @param input
+ * @return input hash as a hex string
+ */
+[[nodiscard]] auto convert_hash_to_hex_string(std::span<unsigned char> input) -> std::string;
+
+/**
+ * Gets the HMAC-SHA256 hash of input with key.
+ * @param input
+ * @param key
+ * @param hash Returns the hashing result.
+ * @return ErrorCode_Success on success.
+ * @return ErrorCode_BadParam if input key exceeds maximum length.
+ * @return ErrorCode_Failure if hash generation fails.
+ */
+[[nodiscard]] auto get_hmac_sha256_hash(
+        std::span<unsigned char const> input,
+        std::span<unsigned char const> key,
+        std::vector<unsigned char>& hash
+) -> ErrorCode;
+
+/**
+ * Gets the SHA256 hash of the input.
+ * @param input
+ * @param hash Returns the hashing result.
+ * @return ErrorCode_Success on success.
+ * @return Same as digest_final_ex and digest_update on failure.
+ * @throw EvpCtxManager::OperationFailed if EvpCtxManager can not be initialized.
```

Comment:
```suggestion
 * @return Same as `digest_final` and `digest_update` on failure.
 * @throw EvpDigestContext::OperationFailed if `EvpDigestContext` cannot be initialized.
```

Although, after you move `EvpDigestContext` into this class, you won't be able to throw its exception, so I would recommend wrapping it in a try-catch and returning some error code.

Reply:
For this and the previous method, we could add something like "See `ERR_get_error` for details about the failure." Otherwise, the failure is quite opaque to the caller.

Reply:
sorry, why "after you move EvpDigestContext into this class, you won't be able to throw its exception"?

Reply:
Sorry, I should rephrase: You might be able to throw it but the caller can't catch that specific type of exception because it's not available in their scope, which would be confusing since this docstring says `EvpDigestContext::OperationFailed` (the best they could do it is catch `TraceableException`, but at that point, I think it's better to just return an error code rather than making the caller catch a very generic exception).

Reply:
sure, but the error_code and docstring for this will also be general because the same error can be returned by multiple callee methods.. Anyway, tried to make a change and updated the docstring.

If they catch a general TraceableException, they can still print the message (with `what()`) which will give them some information, isn't it better than a general error_code?

Reply:
> If they catch a general TraceableException, they can still print the message (with what()) which will give them some information, isn't it better than a general error_code?

That's a fair point. However, that assumes that the caller is only catching the exception to print the error and has no intention of handling the exception. That's probably true in most cases, but isn't really in the spirit of having type-specific exception classes. How about we compromise and create a public `HashUtilsOperationFailed` class and throw that instead of `EvpDigestContext::OperationFailed`?